### PR TITLE
crimson, [WIP]: bring initial support for op tracking

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/osdmap_gate.cc [hunk 1]
```diff
@@ -14,39 +14,84 @@ namespace {
 
 namespace crimson::osd {
 
-void OSDMapGate::OSDMapBlocker::dump_detail(Formatter *f) const
+static constexpr const char* const GATE_NAMES[] = {
```
Comment (tchaikov) [inline src/crimson/osd/osdmap_gate.cc:17]:
```c++
#include <array>
#include <string_view>

using namespace std::literals;

static std::array GATE_NAMES{"OSD"sv, 
                             "PG"sv};
```

File: src/crimson/osd/osdmap_gate.cc [hunk 2]
```diff
@@ -14,39 +14,84 @@ namespace {
...
+  "OSD",
+  "PG"
+};
+
+// prevent the addition of OperationTypeCode-s with no matching OP_NAMES entry:
+static_assert(
+  (sizeof(GATE_NAMES)/sizeof(GATE_NAMES[0])) ==
+  static_cast<int>(OSDMapGateType::last_item));
```
Comment (tchaikov) [inline src/crimson/osd/osdmap_gate.cc:25]:
```c++
static_assert(std::size(GATE_NAMES) == 
              static_cast<int>(OSDMapGateType::last_item));
```

File: src/crimson/admin/osd_admin.cc [hunk 3]
```diff
@@ -188,4 +189,31 @@ class SeastarMetricsHook : public AdminSocketHook {
...
+    op_registry(op_registry)
+  {}
+  seastar::future<tell_result_t> call(const cmdmap_t&,
+				      std::string_view format,
+				      ceph::bufferlist&& input) const final
+  {
+    unique_ptr<Formatter> f{Formatter::create(format, "json-pretty", "json-pretty")};
+    f->open_object_section("ops_in_flight");
+    f->close_section();
+    f->dump_int("num_ops", 0);
```
Comment (tchaikov) [inline src/crimson/admin/osd_admin.cc:210]: might want to `dump_int()` before closing the enclosing section.

File: src/crimson/osd/osd_operation.cc [hunk 4]
```diff
@@ -3,6 +3,13 @@
 
 #include "osd_operation.h"
 #include "common/Formatter.h"
+#include "crimson/common/log.h"
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation.cc:6]: could you fix the typo in the commit message?

File: src/crimson/osd/osd_operation.h [hunk 5]
```diff
@@ -379,12 +381,75 @@ class OperationThrottler : public Blocker,
...
+    static_assert(always_false<T>, "Registry specializarion not found");
+    return {};
+  }
+};
+
+template <class T>
+constexpr static auto EVENT_NAMES = "UnknownEvent";
+
+void dump_event_default(const char name[],
+                        const utime_t& timestamp,
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation.h:410]: i'd suggest use `std::chrono::steady_clock` or `ceph::mono_clock` for this purpose.

File: src/crimson/osd/osd_operation.h [hunk 6]
```diff
@@ -379,12 +381,75 @@ class OperationThrottler : public Blocker,
...
+    utime_t timestamp;
+    void handle(T&, const Operation&) override {
+      timestamp = ceph_clock_now();
+    }
+  } internal_backend;
+
+  // QUESTION: how to convey get_registered_backends() and glue it
+  // with particular Operation? The `events` are supposed to vary.
+  // QUESTION: do we need to handle any other Operation than Client
+  // Request when it comes to Historic?
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation.h:430]: i think we also need to take `RepRequest` and `RecoverySubRequest` into account.

File: src/crimson/osd/osd_operation.h [hunk 7]
```diff
@@ -379,12 +381,75 @@ class OperationThrottler : public Blocker,
...
+  static std::tuple<> get_backends() {
+    static_assert(always_false<T>, "Registry specializarion not found");
+    return {};
+  }
+};
+
+template <class T>
+constexpr static auto EVENT_NAMES = "UnknownEvent";
+
+void dump_event_default(const char name[],
```
Comment (tchaikov) [inline src/crimson/osd/osd_operation.h:409]: hmm, i am not in favor of exposing a symbol with such a general name in the public naming space... probably hide it in a dedicated internal namespace?

**Comment by tchaikov** [comment]
@rzarzynski i am still reviewing this changeset.

**Comment by rzarzynski** [comment]
ACK.

**Comment by tchaikov** [comment]
@rzarzynski the cleanup part is extracted into #41670

**Comment by rzarzynski** [comment]
@rzarzynski: ping.

**Comment by mattbenjamin** [comment]
@rzarzynski what impact does adding support for op tracking have on crimson available iops?

**Comment by rzarzynski** [comment]
@mattbenjamin: haven't profiled yet. My expectation is the penalty should be negligible assuming the extra backends (e.g. LTTng) are disabled at run-time.

Anyway, I'm going to talk with the folks about resurrecting this.

**Comment by rzarzynski** [comment]
Closing in favor of the new PR.

**Pull request closed without merging**